### PR TITLE
librados: invalid free() in rados_getxattrs_next()

### PR DESCRIPTION
**Pull request by guzhongyan**
Invalid free() can cause corruption when getting an object
attribute with empty value.

Check the validity of the pointer before free(). Also move
the free() call at the start of rados_getxattrs_next() to
avoid memory leak.

Fixes: http://tracker.ceph.com/issues/22042
Signed-off-by: Gu Zhongyan <guzhongyan@360.cn>

### CODE REVIEW

File: src/librados/librados.cc [hunk 1]
```diff
@@ -4281,14 +4281,17 @@ extern "C" int rados_getxattrs_next(rados_xattrs_iter_t iter,
 {
   tracepoint(librados, rados_getxattrs_next_enter, iter);
   librados::RadosXattrsIter *it = static_cast<librados::RadosXattrsIter*>(iter);
+  if(it->val) {
```
Comment (liewegas) [inline src/librados/librados.cc]: s/if(/if (/

**Comment by liewegas** [comment]
Also, can you add your attr test case to src/test/pybind/test_rados.py?

File: src/librados/librados.cc (hunk 1 above)
Comment (guzhongyan) [reply src/librados/librados.cc]: Thank you for the review. Done.

**Comment by guzhongyan** [comment]
@liewegas,  attr test case is also added.
with the expaned attr test case:
without my fix, the test case would be failed:
[guzhongyan@rg1-ceph10 ~/ceph/build/bin]$ PYTHONPATH=pybind nosetests ../../src/test/pybind/test_rados.py:TestIoctx.test_obj_xattrs
rados_getxattrs 
*** Error in `/usr/bin/python': double free or corruption (fasttop): 0x0000000002d97800 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x7c619)[0x7f4e43354619]
/lib64/librados.so.2(_ZN8librados15RadosXattrsIterD1Ev+0xd)[0x7f4e374b822d]
/lib64/librados.so.2(rados_getxattrs_end+0x24)[0x7f4e374708d4]
/usr/lib64/python2.7/site-packages/rados.so(+0x174e5)[0x7f4e377884e5]
/lib64/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x6f3d)[0x7f4e440a4cfd]

[guzhongyan@rg1-ceph10 ~/ceph/build/bin]$ PYTHONPATH=pybind nosetests ../../src/test/pybind/test_rados.py:TestIoctx.test_xattrs
rados_getxattrs 
.
----------------------------------------------------------------------
Ran 1 test in 4.299s

OK

After the fix, the result looks good.
[guzhongyan@rg1-ceph10 ~/ceph/build/bin]$ PYTHONPATH=pybind nosetests ../../src/test/pybind/test_rados.py:TestIoctx.test_obj_xattrs
rados_getxattrs 
.
----------------------------------------------------------------------
Ran 1 test in 3.633s

OK

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-02-05_11:48:09-rados-wip-kefu-testing-2018-02-05-1650-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-02-06_06:20:39-rados-wip-kefu-testing-2018-02-05-1650-distro-basic-mira/

**Pull request merged**